src/weighted_list.h

```diff
@@ -0,0 +1,164 @@
+        if(weight >= 0) {
+            for(auto &itr : objects) {
+                if(itr.obj == obj) {
+                    total_weight += (weight - itr.weight);
+                    itr.weight = weight;
+                    return;
+                }
+            }
+
+            // if not found, add to end of list
+            add_item(obj, weight);
+        }
+    }
+
+    /**
+     * This will call the given callback function once for every object in the weighted list.
+     * @param func The callback function.
+     */
+    void apply(std::function<void(const T&)> func) const {
+        for(auto &itr : objects) {
+            func(itr.obj);
+        }
+    }
+
+    /**
+     * This will call the given callback function once for every object in the weighted list.
+     * This is the non-const version.
+     * @param func The callback function.
+     */
+    void apply(std::function<void(T&)> func) {
+        for(auto &itr : objects) {
+            func(itr.obj);
+        }
+    }
+
+    /**
+     * This will return a pointer to an object from the list randomly selected
+     * and biased by weight. If the weighted list is empty or all items in it
+     * have a weight of zero, it will return a NULL pointer.
+     */
+    const T* pick() const {
+        if(total_weight > 0) {
+            return &(objects[pick_ent()].obj);
+        }
+        else {
+            return NULL;
+        }
+    }
+
+    /**
+     * This will return a pointer to an object from the list randomly selected
+     * and biased by weight. If the weighted list is empty or all items in it
+     * have a weight of zero, it will return a NULL pointer. This is the
+     * non-const version so that the returned result may be modified.
+     */
+    T* pick() {
+        if(total_weight > 0) {
+            return &(objects[pick_ent()].obj);
+        }
+        else {
+            return NULL;
+        }
+    }
+
+    /**
+     * This will remove all objects from the list.
+     */
+    void clear() {
+        total_weight = 0;
+        objects.clear();
+    }
+
+    /**
+     * This will return the sum of all the object's weights in the list.
+     */
+    int get_weight() const {
+        return total_weight;
+    }
+
+protected:
+    W total_weight;
+    std::vector<weighted_object<W, T> > objects;
+
+    virtual size_t pick_ent() const =0;
+};
+
+template <typename T> struct weighted_int_list : public weighted_list<int, T> {
+
+protected:
+
+    virtual size_t pick_ent() const {
+        int picked = rng(1, this->total_weight);
+        int accumulated_weight = 0;
+        size_t i;
+        for(i=0; i < this->objects.size(); i++) {
+            accumulated_weight += this->objects[i].weight;
+            if(accumulated_weight >= picked) {
+                break;
+            }
+        }
+        return i;
+    }
+};
+
+template <typename T> struct weighted_float_list : public weighted_list<double, T> {
+
+protected:
+
+    virtual  size_t pick_ent() const {
```

Comment:
This implementation does not work correctly (btw. it should have the `override` keyword and the `virtual` keyword is not needed):

Example: add one entry with weight 0 and another with 100, call `pick_ent`. `total_weight` will be 100, picked (`int`?) can easily be 0, which will trigger the condition `accumulated_weight >= picked` at the first entry already (the entry has weight 0 and should *never* be chosen).

The whole floating point usage has another flaw: the dreaded floating point inaccuracy. Made even worse by using `float` here, but calling `rng_float`, which surprisingly works on `double` and returns a `double`.

Btw.: virtual works through private just fine:
```C++
class A {
private:
    virtual void foo() = 0;
public:
    void call_foo() {
        foo();
    }
};

class B : public A {
private:
    void foo() override {
        ...
    }
};
```